### PR TITLE
[Core] Performance measure framework

### PR DESCRIPTION
**Pull request by erikzaadi**
# Description

Added a way to run the fake integration with configurable total entities count.

This allows us to gather performance metrics.

## Type of change

Please leave one option from the following and delete the rest:

- [x] New feature (non-breaking change which adds functionality)

### CODE REVIEW

File: scripts/smoke-test-base.sh [hunk 1]
```diff
@@ -0,0 +1,35 @@
+#!/usr/bin/env bash
```
Comment (erikzaadi) [inline scripts/smoke-test-base.sh:1]: Remove redundant hashbang

**Review by omby8888 — APPROVED** [review]

File: integrations/fake-integration/fake_org_data/fake_client.py [hunk 2]
```diff
@@ -1,30 +1,55 @@
...
 
 
-fake = Faker()
-
 API_URL = "http://localhost:8000/integration/department"
 USER_AGENT = "Ocean Framework Fake Integration (https://github.com/port-labs/ocean)"
 
+DEFAULT_FAKE_PERSONS_SEED_START = 2
+DEFAULT_FAKE_PERSONS_SEED_END = 19
+BATCH_SIZE = 1000
```
[the remark this answers is not in the document]
Comment (erikzaadi) [reply integrations/fake-integration/fake_org_data/fake_client.py:15]: also add latency seed params

[the remark this answers is not in the document]
Comment (erikzaadi) [reply integrations/fake-integration/fake_org_data/fake_client.py]: Done

File: scripts/smoke-test-base.sh (hunk 1 above)
Comment (erikzaadi) [reply scripts/smoke-test-base.sh:1]: Done

File: integrations/fake-integration/.port/spec.yaml [hunk 3]
```diff
@@ -7,3 +7,29 @@ features:
...
+    default: -1
+  - name: fakePersonsBloatFactor
+    required: false
+    type: integer
+    description: Factor of size to bloat entity (by making the 'bio' string field words length times 60 of factor)
+    default: -1
+  - name: fakePersonsLatencyFactor
+    required: false
+    type: integer
+    description: Factor of size to add latency to each 3rd party API call (if larger than zero, will be a random number between factor and factor times 5)
```
Comment (Tankilevitch) [inline integrations/fake-integration/.port/spec.yaml:34]: can this be generic and not person specific?

File: integrations/fake-integration/.port/spec.yaml [hunk 4]
```diff
@@ -7,3 +7,29 @@ features:
...
+    default: -1
+  - name: fakePersonsBatchSize
+    required: false
+    type: integer
+    description: Batch size of requests against fake 3rd party API
+    default: -1
+  - name: fakePersonsBloatFactor
+    required: false
+    type: integer
+    description: Factor of size to bloat entity (by making the 'bio' string field words length times 60 of factor)
```
Comment (Tankilevitch) [inline integrations/fake-integration/.port/spec.yaml:29]: should be actual size and not bloat. ( raw entity size )

**Review by Tankilevitch — CHANGES_REQUESTED** [review]

File: integrations/fake-integration/fake_org_data/generator.py [hunk 5]
```diff
@@ -8,33 +9,48 @@
 
 fake = Faker()
 
+DEFAULT_AMOUNT = 400
+DEFAULT_BLOAT_FACTOR = 1  # NOTE: Will be multiplied by 1k (approx size of entity json)
```
Comment (Tankilevitch) [inline integrations/fake-integration/fake_org_data/generator.py:13]: lets give it a proper name

**Review by Tankilevitch — CHANGES_REQUESTED** [review]

File: integrations/fake-integration/fake_org_data/generator.py [hunk 6]
```diff
@@ -8,33 +9,48 @@
...
+DEFAULT_LATENCY_FACTOR = (
+    0  # NOTE: Will be multiplied by a random number of factor * 5 if larger than zero
+)
```
[the remark this answers is not in the document]
Comment (erikzaadi) [reply integrations/fake-integration/fake_org_data/generator.py]: Done

File: integrations/fake-integration/.port/spec.yaml (hunk 4 above)
Comment (erikzaadi) [reply integrations/fake-integration/.port/spec.yaml]: Done

File: integrations/fake-integration/fake_org_data/generator.py (hunk 5 above)
Comment (erikzaadi) [reply integrations/fake-integration/fake_org_data/generator.py]: Done

File: .github/workflows/perf-test.yml [hunk 7]
```diff
@@ -0,0 +1,119 @@
...
+          THIRD_PARTY_BATCH_SIZE: 100
+          # THIRD_PARTY_BATCH_SIZE: ${{ inputs.batches }}
+          THIRD_PARTY_LATENCY_FACTOR: 2
+          # THIRD_PARTY_LATENCY_FACTOR: ${{ inputs.latency }}
+          ENTITY_AMOUNT_SEED_START: 100
+          # ENTITY_AMOUNT_SEED_START: ${{ steps.set-perf-params.outputs.SEED_START }}
+          ENTITY_AMOUNT_SEED_END: 150
+          # ENTITY_AMOUNT_SEED_END: ${{ steps.set-perf-params.outputs.SEED_END }}
+          ENTITY_KB_SIZE_FACTOR: 20
+          # ENTITY_KB_SIZE_FACTOR: ${{ inputs.entity_kb_size }}
```
Comment (Tankilevitch) [inline .github/workflows/perf-test.yml:103]: lets use steps variable

File: integrations/_infra/Makefile [hunk 8]
```diff
@@ -11,8 +11,6 @@ define run_checks
 	ruff check . || exit_code=$$?; \
 	echo "Running black"; \
 	black --check . || exit_code=$$?; \
-	echo "Running yamllint"; \
-	yamllint . || exit_code=$$?; \
```
Comment (Tankilevitch) [inline integrations/_infra/Makefile:15]: why?

File: integrations/fake-integration/.port/spec.yaml [hunk 9]
```diff
@@ -7,3 +7,29 @@ features:
...
+  - name: entityAmountSeedEnd
+    required: false
+    type: integer
+    description: End of range for random amount of fake persons per department created (times 5 departments)
+    default: -1
+  - name: entityKbSizeFactor
+    required: false
+    type: integer
+    description: Factor of size of entity (by making the 'bio' string field, ~1k times factor)
+    default: -1
```
Comment (Tankilevitch) [inline integrations/fake-integration/.port/spec.yaml:25]: why not specific size and rather factor?

File: integrations/fake-integration/fake_org_data/fake_client.py [hunk 10]
```diff
@@ -1,30 +1,85 @@
...
+DEFAULT_ENTITY_AMOUNT_SEED_END = 19
+DEFAULT_ENTITY_KB_SIZE_FACTOR = 1
+DEFAULT_THIRD_PARTY_BATCH_SIZE = 1000
+DEFAULT_THIRD_PARTY_LATENCY_FACTOR = 0
+
+ENTITY_AMOUNT_SEED_START_CONFIG_KEY = "entity_amount_seed_start"
+ENTITY_AMOUNT_SEED_END_CONFIG_KEY = "entity_amount_seed_end"
+ENTITY_KB_SIZE_FACTOR_CONFIG_KEY = "entity_kb_size_factor"
+THIRD_PARTY_BATCH_SIZE_CONFIG_KEY = "third_party_batch_size"
+THIRD_PARTY_LATENCY_FACTOR_CONFIG_KEY = "third_party_latency_factor"
```
Comment (Tankilevitch) [inline integrations/fake-integration/fake_org_data/fake_client.py:23]: enum?

File: integrations/fake-integration/fake_org_data/fake_router.py [hunk 11]
```diff
@@ -5,10 +5,15 @@
...
 
 
 def initialize_fake_routes() -> None:
-    @ocean.router.get(FAKE_ROUTE)
-    def get_employees_per_department(department_id: str, limit: int) -> Dict[str, Any]:
-        return generate_fake_persons(department_id, limit)
+    @ocean.router.get(FAKE_ROUTE_BASE)
+    async def get_employees_per_department(
+        department_id: str, limit: int = -1, entity_kb_size: int = -1, latency: int = -1
+    ) -> Dict[str, Any]:
```
Comment (Tankilevitch) [inline integrations/fake-integration/fake_org_data/fake_router.py:15]: lets document what and why its `-1`

File: scripts/run-local-perf-test.sh [hunk 12]
```diff
@@ -0,0 +1,83 @@
+#!/usr/bin/env bash
+
+SCRIPT_BASE="$(cd -P "$(dirname "$0")" && pwd)"
+
+# Usage:
+# run-local-smoke-test.sh
+
```
Comment (Tankilevitch) [inline scripts/run-local-perf-test.sh:7]: rename

File: scripts/run-local-perf-test.sh [hunk 13]
```diff
@@ -0,0 +1,83 @@
...
+# Either have these environment variables set, or change the script below (don't commit that!)
+
+# export PORT_CLIENT_ID=""
+# export PORT_CLIENT_SECRET=""
+# export PORT_BASE_URL=http://localhost:3000
```
Comment (Tankilevitch) [inline scripts/run-local-perf-test.sh:12]: missing all extra params that can be set here

File: scripts/run-local-perf-test.sh [hunk 14]
```diff
@@ -0,0 +1,83 @@
...
+    time "${SCRIPT_BASE}/run-local-smoke-test.sh" | tee "${RUN_LOG_FILE}"
+    END_NS=$(date +%s%N)
+    ELAPSED_MS=$(((END_NS - START_NS) / 1000000))
+    _log "${RUN_TYPE}: Duration $((ELAPSED_MS / 1000)) seconds"
+    echo "Trying to grep"
+    set -x
+    ls -lsa "${RUN_LOG_FILE}"
+    tail -n 15 "${RUN_LOG_FILE}"
+    head -n 5 "${RUN_LOG_FILE}"
+    echo "GREEEEPPPPP"
```
Comment (Tankilevitch) [inline scripts/run-local-perf-test.sh:48]: remove?

File: scripts/run-local-perf-test.sh [hunk 15]
```diff
@@ -0,0 +1,83 @@
...
+    grep -P '\| Upserting' "${RUN_LOG_FILE}" || echo "Failed"
+    set +x
+    _log "${RUN_TYPE}: $(grep '| Upserting' "${RUN_LOG_FILE}" | tail -n 1 | awk -F\| '{print $3};' | xargs)"
+    DELETED=$(grep '| Deleting' "${RUN_LOG_FILE}" | tail -n 1 | awk -F\| '{print $3};' | xargs)
+    if [[ -n "${DELETED}" ]]; then
+        _log "${RUN_TYPE}: ${DELETED}"
+    fi
+
+}
```
Comment (Tankilevitch) [inline scripts/run-local-perf-test.sh]: what is going here? why we are not just writing results to a file and then printing?

Comment (Tankilevitch) [reply scripts/run-local-perf-test.sh]: isn't there a library that can be used to help us

File: integrations/fake-integration/fake_org_data/generator.py [hunk 16]
```diff
@@ -8,33 +9,51 @@
...
 
-def generate_fake_persons(
-    department_id: Union[str, None] = None, amount: Union[int, None] = None
+
+async def generate_fake_persons(
+    department_id: Union[str, None],
+    amount: int,
+    entity_kb_size: int,
+    latency: int,
 ) -> Dict[str, Any]:
```
Comment (Tankilevitch) [inline integrations/fake-integration/fake_org_data/generator.py:24]: lets add description for what are possible values and what each one is used for

**Review by Tankilevitch — CHANGES_REQUESTED** [review]

File: integrations/_infra/Makefile (hunk 8 above)
Comment (erikzaadi) [reply integrations/_infra/Makefile:15]: yamlint isn't availble at integrations, only in main project

File: scripts/run-local-perf-test.sh (hunk 12 above)
Comment (erikzaadi) [reply scripts/run-local-perf-test.sh:7]: to?

File: scripts/run-local-perf-test.sh (hunk 14 above)
Comment (erikzaadi) [reply scripts/run-local-perf-test.sh:48]: Part of current debugging efforts, won't be in the released

File: scripts/run-local-perf-test.sh (hunk 12 above)
Comment (erikzaadi) [reply scripts/run-local-perf-test.sh]: done

File: integrations/fake-integration/fake_org_data/fake_router.py (hunk 11 above)
Comment (erikzaadi) [reply integrations/fake-integration/fake_org_data/fake_router.py]: done

File: scripts/run-local-perf-test.sh (hunk 13 above)
Comment (erikzaadi) [reply scripts/run-local-perf-test.sh:12]: Added

File: scripts/run-local-perf-test.sh (hunk 15 above)
Comment (erikzaadi) [reply scripts/run-local-perf-test.sh]: Managed to make it work with ruby

File: integrations/fake-integration/fake_org_data/generator.py (hunk 16 above)
Comment (erikzaadi) [reply integrations/fake-integration/fake_org_data/generator.py]: Added

File: integrations/fake-integration/fake_org_data/fake_client.py (hunk 10 above)
Comment (erikzaadi) [reply integrations/fake-integration/fake_org_data/fake_client.py]: Done

File: integrations/_infra/Makefile (hunk 8 above)
Comment (erikzaadi) [reply integrations/_infra/Makefile:15]: Didn't exist before (see https://github.com/port-labs/ocean/commit/bdb4dd7febfe46a18c032545e46222109c51ace5#diff-53118e2f6d6f81647184d5bc94a5275ba819e5d999cee17a0e022b8d54541158 and parents) adding a ticket [PORT-11773](https://getport.atlassian.net/browse/PORT-11773)

File: .github/workflows/perf-test.yml [hunk 17]
```diff
@@ -0,0 +1,104 @@
...
+        default: "0"
+        options:
+          - "0"
+          - "5"
+          - "10"
+          - "25"
+          - "100"
+      amount:
+        type: choice
+        description: Explicit amount of entities created (multiplied by 5 which is the amount of departments)
```
[the remark this answers is not in the document]
Comment (erikzaadi) [reply .github/workflows/perf-test.yml]: Ignored explicitly in perf, see https://github.com/port-labs/ocean/pull/1176/files#diff-c97bd6cc14678e298598211aec6f93a9cbccc71abd2e9df307d832b36cefc69bR31

File: .github/workflows/perf-test.yml [hunk 18]
```diff
@@ -0,0 +1,104 @@
...
+        options:
+          - "1"
+          - "5"
+          - "10"
+          - "25"
+          - "100"
+          - "500"
+      latency:
+        type: choice
+        description: Factor of size to add latency to each 3rd party API call (if larger than zero, will be a random number between factor and factor times 5)
```
[the remark this answers is not in the document]
Comment (erikzaadi) [reply .github/workflows/perf-test.yml]: Removed factor made all explicit

File: .github/workflows/perf-test.yml [hunk 19]
```diff
@@ -0,0 +1,104 @@
...
+          SMOKE_TEST_SUFFIX: ${{ github.run_id }}
+          # THIRD_PARTY_BATCH_SIZE: ${{ inputs.batches }}
+          THIRD_PARTY_BATCH_SIZE: 100
+          # THIRD_PARTY_LATENCY_FACTOR: ${{ inputs.latency }}
+          THIRD_PARTY_LATENCY_FACTOR: 0
+          ENTITY_AMOUNT_SEED_START: 500
+          # ENTITY_AMOUNT_SEED_START: ${{ inputs.amount }}
+          ENTITY_AMOUNT_SEED_END: 500
+          # ENTITY_AMOUNT_SEED_END: ${{ inputs.amount }}
+          # ENTITY_KB_SIZE: ${{ inputs.entity_kb_size }}
```
[the remark this answers is not in the document]
Comment (erikzaadi) [reply .github/workflows/perf-test.yml]: Will use the inputs when merged, moved the commented ones to the `[DROP]` commit to make it easier to review

File: integrations/fake-integration/.port/spec.yaml [hunk 20]
```diff
@@ -7,3 +7,29 @@ features:
...
+  - name: thirdPartyBatchSize
+    required: false
+    type: integer
+    description: Batch size of requests against fake 3rd party API
+    default: -1
+  - name: thirdPartyLatencyFactor
+    required: false
+    type: integer
+    description: Factor of size to add latency to each 3rd party API call (if larger than zero, will be a random number between factor and factor times 5)
```
[the remark this answers is not in the document]
Comment (erikzaadi) [reply integrations/fake-integration/.port/spec.yaml]: Changed to explicit ms

**Review by Tankilevitch — CHANGES_REQUESTED** [review]

File: scripts/run-smoke-test.sh [hunk 21]
```diff
@@ -50,13 +36,20 @@ echo "Found release ${TAR_FILE}, triggering fake integration with ID: '${INTEGRA
 docker run --rm -i \
     --entrypoint 'bash' \
     -v "${TAR_FULL_PATH}:/opt/dist/${TAR_FILE}" \
-    -v "${TEMP_DIR}/resources:/app/.port/resources" \
+    -v "${TEMP_RESOURCES_DIR}:/opt/port-resources" \
```
[the remark this answers is not in the document]
Comment (erikzaadi) [reply scripts/run-smoke-test.sh:39]: I used Shalev's new resource dir env var, makes it cleaner and possible to run locally as well

**Review by Tankilevitch — APPROVED** [review]
✅

**Pull request merged**